`{Release Hash}` Renaming Token

Alanoll opened the pull request:
#### Description
Implementing the `{Release Hash}` for use in renaming.

The parsed Release Hash is stored in the DB table `EpisodeFile`.

#### Database Migration
YES - 204


#### Issues Fixed or Closed by this PR
* Closes #6570

Alanoll commented on src/NzbDrone.Core/Datastore/Migration/203_add_release_hash.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace NzbDrone.Core.Datastore.Migration
+{
+    [Migration(203)]
```
I wasn't sure how DB migrations are kept in sync across forks, so I just incremented to the next ID.

markus101 replied:
203 is already used in https://github.com/Sonarr/Sonarr/pull/6567 will need to be bumped to 204

Alanoll commented on src/NzbDrone.Core/Parser/Parser.cs:
I tweaked the RegEx's to open them up alittle more to get any trailing Hashes in the strings, and allow for parenthesis as well. Half of them tried to do end string, but also had a "stop at first period" assuming the period would be the period of the file extension which isn't always the case.

All the unit tests passed, and I added more for parenthesized hashes.

markus101 commented on src/NzbDrone.Core.Test/OrganizerTests/FileNameBuilderTests/FileNameBuilderFixture.cs:
```diff
@@ -991,6 +991,67 @@ public void should_use_updated_media_info_if_token_configured_and_revision_is_ol
+        [TestCase("Series Title - 031 - The Resolution to Kill [Lunar].avi", "")]
+        [TestCase("[ACX]Series Title 01 Episode Name [Kosaka] [9C57891E].mkv", "9C57891E")]
+        [TestCase("[S-T-D] Series Title! - 06 (1280x720 10bit AAC) [59B3F2EA].mkv", "59B3F2EA")]
+        public void should_replace_release_space_hash_with_scenename_hash(string sceneName, string hash)
```
```suggestion
        public void should_replace_release_hash_with_scenename_hash(string sceneName, string hash)
```

markus101 commented on src/NzbDrone.Core.Test/OrganizerTests/FileNameBuilderTests/FileNameBuilderFixture.cs:
```diff
@@ -991,6 +991,67 @@ public void should_use_updated_media_info_if_token_configured_and_revision_is_ol
+        [TestCase("Series Title - 031 - The Resolution to Kill [Lunar].avi", "")]
+        [TestCase("[ACX]Series Title 01 Episode Name [Kosaka] [9C57891E].mkv", "9C57891E")]
+        [TestCase("[S-T-D] Series Title! - 06 (1280x720 10bit AAC) [59B3F2EA].mkv", "59B3F2EA")]
+        public void should_replace_release_space_hash_with_filename_hash(string fileName, string hash)
```
```suggestion
        public void should_replace_release_hash_with_filename_hash(string fileName, string hash)
```

markus101 commented on src/NzbDrone.Core/Organizer/FileNameBuilder.cs:
```diff
@@ -1101,6 +1102,27 @@ private string GetOriginalFileName(EpisodeFile episodeFile, bool useCurrentFilen
             return Path.GetFileNameWithoutExtension(episodeFile.RelativePath);
         }
 
+        private string GetReleaseHash(EpisodeFile episodeFile)
```
This logic can all be done in the migration, then it doesn't need to be done at the time of naming. You could use what this PR does as an example. https://github.com/Sonarr/Sonarr/pull/656

Alanoll replied:
The db migration now performs the parsing. 
https://github.com/Sonarr/Sonarr/pull/6567/files#r1507044123

The timeout y'all discussed in the other PR may be an issue here as well, since it's also reparsing all titles.

Alanoll commented:
- Updated the DB migration to perform the parsing, and incremented it's ID to 204.
- remove the Unit Tests that were testing the title parsing within the FileNameBuilder
- added a Unit Test to ensure null/empty Release Hashes return empty string correctly

Alanoll commented:
Realized this morning that I forgot to add the new token to the File Naming Tokens dialog. Added it to the dialog and shown only for Anime patterns.

markus101 commented on src/NzbDrone.Core/Datastore/Migration/204_add_release_hash.cs:
```diff
@@ -0,0 +1,75 @@
+                while (reader.Read())
+                {
+                    var id = reader.GetInt32(0);
+                    var sceneName = reader.IsDBNull(1) ? null : reader.GetString(1);
+                    var relativePath = reader.IsDBNull(2) ? null : reader.GetString(2);
+                    var originalFilePath = reader.IsDBNull(3) ? null : reader.GetString(3);
```
These can be simplified to `var x= reader[n] as string;` which correctly handles null values.

Alanoll replied:
Good tip. I'm so ingrained in the older pattern, never thought of using that.

markus101 commented on src/NzbDrone.Core/Datastore/Migration/204_add_release_hash.cs:
```diff
@@ -0,0 +1,75 @@
+                    {
+                        title = sceneName;
+                    }
+                    else if (relativePath.IsNotNullOrWhiteSpace())
```
This will never be null. so the `originalFilePath` branch will never be hit.

Alanoll replied:
Obsolete

markus101 commented on src/NzbDrone.Core/Datastore/Migration/204_add_release_hash.cs:
```diff
@@ -0,0 +1,75 @@
+                        title = Path.GetFileNameWithoutExtension(originalFilePath);
+                    }
+
+                    if (!title.IsNullOrWhiteSpace())
```
Since the relative path probably doesn't have the hash (given it was only available via a Preferred Word match), I think this should be two tiered.

1. Check the scene name or original file name, giving preference to the scene name as above if either is available
2. Check the existing file name, which will cover cases where the original information is unavailable, but the file name does have it.

Something like: 

```
ParsedEpisodeInfo parsedEpisodeInfo = null;

var originalTitle = sceneName ?? originalFilePath.IsNotNullOrWhiteSpace() ? Path.GetFileNameWithoutExtension(originalFilePath) : null;

if (originalTitle.IsNotNullOrWhiteSpace())
{
  parsedEpisodeInfo = Parser.Parser.ParseTitle(originalTitle);
}

if ( parsedEpisodeInfo?.ReleaseHash.IsNullOrWhiteSpace())
{
  parseEpisodeInfo = Parser.Parser.ParseTitle(Path.GetFileNameWithoutExtension(relativePath));
}

if (parsedEpisodeInfo?.ReleaseHash.IsNotNullOrWhiteSpace())
{
  updates.Add(new
	{
	    Id = id,
	    ReleaseHash = parsedEpisodeInfo.ReleaseHash
	});
}
```

This way it's best effort to get the value and at most will parse 2x per file.

Alanoll replied:
Implemented the suggested changes, but was explicit instead of the null coalescing operator in case `sceneName` somehow is an empty string since FileNameBuilder performs the same check

Alanoll commented on src/NzbDrone.Core/Datastore/Migration/204_add_release_hash.cs:
```diff
@@ -0,0 +1,75 @@
+            using (var cmd = conn.CreateCommand())
+            {
+                cmd.Transaction = tran;
+                cmd.CommandText = "SELECT \"Id\", \"SceneName\", \"RelativePath\", \"OriginalFilePath\" FROM \"EpisodeFiles\" WHERE \"ReleaseHash\" IS NULL";
```
Removed the condition

markus101 commented on src/NzbDrone.Core/Datastore/Migration/204_add_release_hash.cs:
```diff
@@ -0,0 +1,71 @@
+
+                    ParsedEpisodeInfo parsedEpisodeInfo = null;
+
+                    var originalTitle = sceneName.IsNotNullOrWhiteSpace() ? sceneName : (originalFilePath.IsNotNullOrWhiteSpace() ? Path.GetFileNameWithoutExtension(originalFilePath) : null);
```
Since `sceneName` is null when it's not set this can be shortened to:

```suggestion
                    var originalTitle = sceneName ?? originalFilePath.IsNotNullOrWhiteSpace() ? Path.GetFileNameWithoutExtension(originalFilePath) : null;
```

Alanoll replied:
Can SceneName ever be the empty string in the database?
I went the IsNullOrWhiteSpace route since that's what FileNameBuilder did.

markus101 replied:
Should only be only be null or a valid value unless someone manually altered their DB to set it to empty string.

Yeah, we use `Is(Not)NullOrWhiteSpace` across the app to avoid having to check bot, but in the migration just the null check is fine.

markus101 replied:
Thinking about it a bit more, not much harm in the empty string check, so this would be what I'd do to avoid the double ternary.

```suggestion
                    var originalTitle = sceneName;
                    
                     if (originalTitle.IsNullOrWhiteSpace() && originalFilePath.IsNotNullOrWhiteSpace())
                     {
                         originalTitle = Path.GetFileNameWithoutExtension(originalFilePath);
                     }
```

Alanoll commented on src/NzbDrone.Core/Parser/Parser.cs:
```diff
@@ -492,7 +492,7 @@ public static class Parser
         private static readonly Regex FileExtensionRegex = new Regex(@"\.[a-z0-9]{2,4}$",
                                                                 RegexOptions.IgnoreCase | RegexOptions.Compiled);
 
-        private static readonly RegexReplace SimpleTitleRegex = new RegexReplace(@"(?:(480|540|576|720|1080|2160)[ip]|[xh][\W_]?26[45]|DD\W?5\W1|[<>?*]|848x480|1280x720|1920x1080|3840x2160|4096x2160|(8|10)b(it)?|10-bit)\s*?",
```
This was causing any hash value with either `8B` or `10B` to be removed when the title is simplified. For example `[4228bb8c]` would become `[422b8c]` and then failed the hash regex because it's not 8 characters.

The updated expression should still match `8b`, `8bit`, `10b`, `10bit`, or `10-bit` for removal, but `8b` and `10b` can't have an immediate word character after it. I didn't use `\w` because of the underscore and `_8b_` could be what people use in their files.

Alanoll replied:
There's still cases where the hash won't match, but I think a decision would need to be made.
Hashes that end with `10b` or `8b` still might not match e.g. `[15AF298B]` or `[15AF210B]`

Alanoll replied:
An alternative could be to check if the `<hash>` match has content, and if not, do an explicit RegEx against the raw release title trying to find the hash.

Pros:
- Don't have to keep tweaking existing expressions for niche cases of Anime naming
- Could find hashes in non-anime releases if they had them (I don't know of any off the top of my head)

Con:
- Increases processing time with an additional check. The check is only perform when one of the title matches hits, so it's not a 2x across the board. It would be a simple regex `(?<hash>[(\[][0-9a-f]{8}[)\]])$`, so should be relatively quick.
- Could find hashes in non-anime releases (leading to questions of "why does this one have a hash but that one doesn't?")

markus101 replied:
I think you should be able to do a lookbehind and only match if it's proceeded by a space, underscore, or period, or a negative look behind to make sure it's not proceeded by a valid character in the hash.

markus101 commented on src/NzbDrone.Core/Parser/Parser.cs:
```diff
@@ -492,7 +492,7 @@ public static class Parser
                                                                 RegexOptions.IgnoreCase | RegexOptions.Compiled);
 
-        private static readonly RegexReplace SimpleTitleRegex = new RegexReplace(@"(?:(480|540|576|720|1080|2160)[ip]|[xh][\W_]?26[45]|DD\W?5\W1|[<>?*]|848x480|1280x720|1920x1080|3840x2160|4096x2160|(8|10)b(it)?|10-bit)\s*?",
+        private static readonly RegexReplace SimpleTitleRegex = new RegexReplace(@"(?:(480|540|576|720|1080|2160)[ip]|[xh][\W_]?26[45]|DD\W?5\W1|[<>?*]|848x480|1280x720|1920x1080|3840x2160|4096x2160|(8|10)(b(?![a-z0-9])|bit)|10-bit)\s*?",
```
```suggestion
        private static readonly RegexReplace SimpleTitleRegex = new RegexReplace(@"(?:(480|540|576|720|1080|2160)[ip]|[xh][\W_]?26[45]|DD\W?5\W1|[<>?*]|848x480|1280x720|1920x1080|3840x2160|4096x2160|(?<![]a-f0-9])(8|10)b(it)?|10-bit)\s*?",
```

This works with the example title provided.

Alanoll replied:
I combined our two patterns. Your look behind, but with my look ahead.

Some mocked up example parses:
```
[DHD] Series Title! - 01 - Episode Title [1080p][10bit][15AF298B]
[DHD] Series Title! - 01 - Episode Title [1080p 10bit][15AF298B]
[DHD] Series Title! - 01 - Episode Title [1080p 10b][15AF298B]
[DHD] Series Title! - 01 - Episode Title [1080p][10b][15AF298B]
[DHD] Series Title! - 01 - Episode Title [1080p 10b][8B15AF29]
[DHD] Series Title! - 01 - Episode Title [1080p][10b][8B15AF29]
[DHD] Series Title! - 01 - Episode Title [1080p][8bit][15AF298B]
[DHD] Series Title! - 01 - Episode Title [1080p 8bit][15AF298B]
[DHD] Series Title! - 01 - Episode Title [1080p 8b][15AF298B]
[DHD] Series Title! - 01 - Episode Title [1080p][8b][15AF298B]
[DHD] Series Title! - 01 - Episode Title [1080p 8b][8B15AF29]
[DHD] Series Title! - 01 - Episode Title [1080p][8b][8B15AF29]
```

With just the suggested change:
![image](https://github.com/Sonarr/Sonarr/assets/21229471/10535d49-7a23-48f7-b1b1-521e033bd854)

With the one I committed:
![image](https://github.com/Sonarr/Sonarr/assets/21229471/178fc38a-2094-43fb-99c4-5eb00ca90f95)

markus101 replied:
Currently only tests for `8b` and `10b` are covered, would be good to add additional ones.

Alanoll replied:
What additional ones where you thinking of?
I split off the explicit 8b and 10b within the hashes to another function, with begins with, ends with, and contains variants.

markus101 replied:
Perfect, I think I just missed them as I reviewed.

markus101 commented on src/NzbDrone.Core.Test/ParserTests/AnimeMetadataParserFixture.cs:
```diff
@@ -22,12 +22,39 @@ public class AnimeMetadataParserFixture : CoreTest
+        [TestCase("[S-T-D] Series Title! - 06 (1280x720 10bit AAC) (59B3F2EA).mkv", "S-T-D", "59B3F2EA")]
+        [TestCase("[DHD] Series Title! - 08 (1280x720 10bit AAC) (8B8BF2EA).mkv", "DHD", "8B8BF2EA")]
+        [TestCase("[DHD] Series Title! - 10 (1280x720 10bit AAC) (10BBF2EA).mkv", "DHD", "10BBF2EA")]
+        public void should_parse_absolute_numbers_parenthesized_hash(string postTitle, string subGroup, string hash)
```
No need to create a duplicate function with many duplicate tests, add the relevant tests to `should_parse_absolute_numbers`.

Alanoll replied:
I recombined them to `should_parse_absolute_numbers` but also renamed it to reflect what is being tested.

Alanoll commented on src/NzbDrone.Core/Organizer/FileNameBuilder.cs:
```diff
@@ -623,6 +623,7 @@ private void AddEpisodeFileTokens(Dictionary<string, Func<TokenMatch, string>> t
             tokenHandlers["{Original Title}"] = m => GetOriginalTitle(episodeFile, useCurrentFilenameAsFallback);
             tokenHandlers["{Original Filename}"] = m => GetOriginalFileName(episodeFile, useCurrentFilenameAsFallback);
             tokenHandlers["{Release Group}"] = m => Truncate(episodeFile.ReleaseGroup, m.CustomFormat) ?? m.DefaultValue("Sonarr");
+            tokenHandlers["{Release Hash}"] = m => GetReleaseHash(episodeFile);
```
Done

Alanoll commented on src/NzbDrone.Core/MediaFiles/EpisodeImport/ImportApprovedEpisodes.cs:
```diff
@@ -95,6 +95,7 @@ public List<ImportResult> Import(List<ImportDecision> decisions, bool newDownloa
                     episodeFile.SeasonNumber = localEpisode.SeasonNumber;
                     episodeFile.Episodes = localEpisode.Episodes;
                     episodeFile.ReleaseGroup = localEpisode.ReleaseGroup;
+                    episodeFile.ReleaseHash = localEpisode.FileEpisodeInfo?.ReleaseHash;
```
Implemented.

Alanoll commented:
Rebased onto develop to clear merge conflict from #6608
